### ISSUE TITLE
MultiboxLoss data format?

### ISSUE THREAD

**Issue by tachim2** [xiaohaoChen/rrc_detection#3]
What must the shapes of the label tensor be when writing a custom layer? I have a `PythonLayer` that I'm using to load data from my custom data sources, but I'm not sure what shapes to make the label tensor. Currently I'm trying to set the shape to be (num_boxes, 4), but that's throwing an error because num_boxes != 1.

**Comment by tachim2** [comment]
Hmm, looks like it should be `(1, 1, <num boxes>, 8)` though not sure what the 8 columns are!

**Comment by tachim2** [comment]
Is the idea to pre-assign the boxes to the priors before passing them to `MultiBoxLoss`? Is there a way to do that in python?

**Comment by tachim2** [comment]
This code finally runs, but yields only 0 loss so I'm guessing something went wrong.

This is the `forward(self, bottom, top)` function:
```
        rgb = scipy.misc.imresize(self.rgb, self.image_size)
        top[0].data[0, ...] = rgb.transpose((2, 1, 0)) 
        top[1].data[0, 0, :, :4] = self.bbox  # [n x 4] array where each column is the fraction offset of the box along the axis, (xmin, ymin, xmax, ymax)
        top[1].data[0, 0, :, 4] = 0 # background?
        top[1].data[0, 0, :, 5] = 1 # car?
```

where the .prototxt looks like

```
name: "VGG_KITTI_RRC_2560x768_kitti_car_test"     
layer {     
  name: "data"     
  type: "Python"     
  top: "data"     
  top: "label"     
  python_param {     
    module: "caffe_layer"     
    layer: "DataLayer"
    param_str: "{\"phase\": \"test\", \"size\": [768, 2560]}"     
  }     
}           
layer {               
  name: "conv1_1"         
  type: "Convolution" 
  ...
```

The training output looks like
```
I0421 02:54:22.471334   976 solver.cpp:231] Iteration 0, loss = 0
I0421 02:54:22.471360   976 solver.cpp:247]     Train net output #0: mbox_loss = 0 (* 1 = 0 loss)
I0421 02:54:22.471367   976 solver.cpp:247]     Train net output #1: mbox_loss2 = 0 (* 1 = 0 loss)
I0421 02:54:22.471374   976 solver.cpp:247]     Train net output #2: mbox_loss3 = 0 (* 1 = 0 loss)
I0421 02:54:22.471377   976 solver.cpp:247]     Train net output #3: mbox_loss4 = 0 (* 1 = 0 loss)
I0421 02:54:22.471384   976 solver.cpp:247]     Train net output #4: mbox_loss5 = 0 (* 1 = 0 loss)
I0421 02:54:22.471397   976 sgd_solver.cpp:106] Iteration 0, lr = 0.0005
I0421 02:55:28.848285   976 solver.cpp:231] Iteration 10, loss = 0
I0421 02:55:28.848400   976 solver.cpp:247]     Train net output #0: mbox_loss = 0 (* 1 = 0 loss)
I0421 02:55:28.848407   976 solver.cpp:247]     Train net output #1: mbox_loss2 = 0 (* 1 = 0 loss)
I0421 02:55:28.848410   976 solver.cpp:247]     Train net output #2: mbox_loss3 = 0 (* 1 = 0 loss)
I0421 02:55:28.848413   976 solver.cpp:247]     Train net output #3: mbox_loss4 = 0 (* 1 = 0 loss)
I0421 02:55:28.848417   976 solver.cpp:247]     Train net output #4: mbox_loss5 = 0 (* 1 = 0 loss)
I0421 02:55:28.848420   976 sgd_solver.cpp:106] Iteration 10, lr = 0.0005
I0421 02:56:32.323758   976 solver.cpp:231] Iteration 20, loss = 0
I0421 02:56:32.323890   976 solver.cpp:247]     Train net output #0: mbox_loss = 0 (* 1 = 0 loss)
I0421 02:56:32.323896   976 solver.cpp:247]     Train net output #1: mbox_loss2 = 0 (* 1 = 0 loss)
I0421 02:56:32.323900   976 solver.cpp:247]     Train net output #2: mbox_loss3 = 0 (* 1 = 0 loss)
I0421 02:56:32.323904   976 solver.cpp:247]     Train net output #3: mbox_loss4 = 0 (* 1 = 0 loss)
I0421 02:56:32.323907   976 solver.cpp:247]     Train net output #4: mbox_loss5 = 0 (* 1 = 0 loss)
```

My guess is something silly is going on here -- any idea?

Edit: I modified the bbox format to follow `rrc_test.py`, still no luck:
```
top[0].data[0, ...] = rgb.transpose((2, 1, 0))
top[1].data[0, 0, :, 1] = 1
top[1].data[0, 0, :, 2] = 1
top[1].data[0, 0, :, 3:7] = self.bbox
```

Edit 2: looked at MultiboxLossLayer.cpp, seems like 
```
top[1].data[0, 0, :, 0] = range(len(self.bbox))
top[1].data[0, 0, :, 1] = 1
top[1].data[0, 0, :, 3:7] = self.bbox
top[1].data[0, 0, :, 7] = 0
``` 
is on the right track.

Edit 3: it seems the bounding boxes must be normalized to lie in the range [0, 1], but even so the loss becomes nan after a few iterations:

```
I0421 03:41:26.762118  1244 solver.cpp:282] Solving VGG_KITTI_RRC_768x512_kitti_car_train
I0421 03:41:26.762121  1244 solver.cpp:283] Learning Rate Policy: step
I0421 03:41:27.867493  1244 solver.cpp:231] Iteration 0, loss = 131.806
I0421 03:41:27.867522  1244 solver.cpp:247]     Train net output #0: mbox_loss = 50.3912 (* 1 = 50.3912 loss)
I0421 03:41:27.867527  1244 solver.cpp:247]     Train net output #1: mbox_loss2 = 41.1194 (* 1 = 41.1194 loss)
I0421 03:41:27.867530  1244 solver.cpp:247]     Train net output #2: mbox_loss3 = 38.0257 (* 1 = 38.0257 loss)
I0421 03:41:27.867534  1244 solver.cpp:247]     Train net output #3: mbox_loss4 = 35.16 (* 1 = 35.16 loss)
I0421 03:41:27.867538  1244 solver.cpp:247]     Train net output #4: mbox_loss5 = 33.8954 (* 1 = 33.8954 loss)
I0421 03:41:27.867548  1244 sgd_solver.cpp:106] Iteration 0, lr = 0.0005
I0421 03:41:38.310899  1244 solver.cpp:231] Iteration 10, loss = 89.8108
I0421 03:41:38.310930  1244 solver.cpp:247]     Train net output #0: mbox_loss = 78.38 (* 1 = 78.38 loss)
I0421 03:41:38.310936  1244 solver.cpp:247]     Train net output #1: mbox_loss2 = 41.9339 (* 1 = 41.9339 loss)
I0421 03:41:38.310941  1244 solver.cpp:247]     Train net output #2: mbox_loss3 = 32.4417 (* 1 = 32.4417 loss)
I0421 03:41:38.310946  1244 solver.cpp:247]     Train net output #3: mbox_loss4 = 19.596 (* 1 = 19.596 loss)
I0421 03:41:38.310951  1244 solver.cpp:247]     Train net output #4: mbox_loss5 = 17.5602 (* 1 = 17.5602 loss)
I0421 03:41:38.310956  1244 sgd_solver.cpp:106] Iteration 10, lr = 0.0005
I0421 03:41:48.347851  1244 solver.cpp:231] Iteration 20, loss = nan
I0421 03:41:48.347877  1244 solver.cpp:247]     Train net output #0: mbox_loss = nan (* 1 = nan loss)
I0421 03:41:48.347882  1244 solver.cpp:247]     Train net output #1: mbox_loss2 = nan (* 1 = nan loss)
I0421 03:41:48.347885  1244 solver.cpp:247]     Train net output #2: mbox_loss3 = nan (* 1 = nan loss)
I0421 03:41:48.347889  1244 solver.cpp:247]     Train net output #3: mbox_loss4 = nan (* 1 = nan loss)
I0421 03:41:48.347892  1244 solver.cpp:247]     Train net output #4: mbox_loss5 = nan (* 1 = nan loss)
I0421 03:41:48.347895  1244 sgd_solver.cpp:106] Iteration 20, lr = 0.0005
```

Edit 4: this finally appears to train after reducing the learning rate (noticed a comment about this in the code, since I'm using a batch size of 1 on 1 GPU). Can you give any guidance on what the loss curve looks like on your end?

**Comment by xiaohaoChen** [comment]
Look at  annotated_data_layer.cpp: line 214
```shell
              top_label[idx++] = item_id;      // the id of the images in the batch 
              top_label[idx++] = anno_group.group_label();   // the label id of the object
              top_label[idx++] = anno.instance_id();       // instance id of the object in the image 
              top_label[idx++] = bbox.xmin();    // the box should be normalized
              top_label[idx++] = bbox.ymin();
              top_label[idx++] = bbox.xmax();
              top_label[idx++] = bbox.ymax();
              top_label[idx++] = bbox.difficult();    //you can set it 0 defaultly
```
But I think converting your labels to VOC type and use the proviced scripts to generate lmdb for training is a better idea, or some other bugs may occur.

If the loss becomes nan again in your setting, you can reduce the learning to 0.00005.

**Comment by tachim2** [comment]
Thank you! indeed the label id vs instance id was not correct.
